amazon/ion/ioncmodule.c

```diff
@@ -0,0 +1,1369 @@
+        }
+        Py_DECREF(py_sid);
+    }
+    else {
+        ION_STRING string_value;
+        ion_string_from_py(symbol_text, &string_value);
+        if (is_value) {
+            err = ion_writer_write_symbol(writer, &string_value);
+        }
+        else {
+            err = ion_writer_add_annotation(writer, &string_value);
+        }
+    }
+    Py_DECREF(symbol_text);
+    IONCHECK(err);
+    iRETURN;
+}
+
+/*
+ *  Writes annotations
+ *
+ *  Args:
+ *      writer:  An ion writer
+ *      obj: A sequence of ion python annotations
+ *
+ */
+static iERR ionc_write_annotations(hWRITER writer, PyObject* obj) {
+    iENTER;
+    PyObject* annotations = NULL;
+    if (PyObject_HasAttrString(obj, "ion_annotations")) {
+        annotations = PyObject_GetAttrString(obj, "ion_annotations");
+    }
+
+    if (annotations == NULL || PyObject_Not(annotations)) SUCCEED();
+
+    annotations = PySequence_Fast(annotations, "expected sequence");
+    Py_ssize_t len = PySequence_Size(annotations);
+    Py_ssize_t i;
+
+    for (i = 0; i < len; i++) {
+        PyObject* pyAnnotation = PySequence_Fast_GET_ITEM(annotations, i);
+        Py_INCREF(pyAnnotation);
+        if (PyUnicode_Check(pyAnnotation)) {
+            ION_STRING annotation;
+            ion_string_from_py(pyAnnotation, &annotation);
+            err = ion_writer_add_annotation(writer, &annotation);
+        }
+        else if (PyObject_TypeCheck(pyAnnotation, (PyTypeObject*)_py_symboltoken_constructor)){
+            err = ionc_write_symboltoken(writer, pyAnnotation, /*is_value=*/FALSE);
+        }
+        Py_DECREF(pyAnnotation);
+        if (err) break;
+    }
+    Py_XDECREF(annotations);
+fail:
+    Py_XDECREF(annotations);
+    cRETURN;
+}
+
+/*
+ *  Writes a list or a sexp
+ *
+ *  Args:
+ *      writer:  An ion writer
+ *      sequence: An ion python list or sexp
+ *      tuple_as_sexp: Decides if a tuple is treated as sexp
+ *
+ */
+static iERR ionc_write_sequence(hWRITER writer, PyObject* sequence, PyObject* tuple_as_sexp) {
+    iENTER;
+    PyObject* child_obj = NULL;
+    sequence = PySequence_Fast(sequence, "expected sequence");
+    Py_ssize_t len = PySequence_Size(sequence);
+    Py_ssize_t i;
+
+    for (i = 0; i < len; i++) {
+        child_obj = PySequence_Fast_GET_ITEM(sequence, i);
+        Py_INCREF(child_obj);
+
+        IONCHECK(Py_EnterRecursiveCall(" while writing an Ion sequence"));
+        err = ionc_write_value(writer, child_obj, tuple_as_sexp);
+        Py_LeaveRecursiveCall();
+        IONCHECK(err);
+
+        Py_DECREF(child_obj);
+        child_obj = NULL;
+    }
+fail:
+    Py_XDECREF(child_obj);
+    Py_DECREF(sequence);
+    cRETURN;
+}
+
+/*
+ *  Writes a struct
+ *
+ *  Args:
+ *      writer:  An ion writer
+ *      map: An ion python struct
+ *      tuple_as_sexp: Decides if a tuple is treated as sexp
+ *
+ */
+static iERR ionc_write_struct(hWRITER writer, PyObject* map, PyObject* tuple_as_sexp) {
```

Comment:
Before we use python dict for ion struct which didn't support duplicate keys, I re-write this api.